File: tcping/ConsoleIO.m
//
//  ConsoleIO.m
//  tcping
//
//  Created by ParadiseDuo on 2020/5/11.
//  Copyright © 2020 MacClient. All rights reserved.
//

#import "ConsoleIO.h"
#import "Version.h"

@implementation ConsoleIO
+ (void)writeMessage:(NSString *)message to:(NSInteger)to {
    switch (to) {
        case OutputTypeError:
            NSLog(@"%@", message);
            break;
        default:
            fputs(message.UTF8String, stderr);
            break;
    }
}

+ (void)printUsage {
   
    NSString * v = @""
"\n      dP                     oo"
"\n      88"
"\n    d8888P .d8888b. 88d888b. dP 88d888b. .d8888b.    {Version: %@}"
"\n      88   88       88    88 88 88    88 88    88"
"\n      88   88.      88.  .88 88 88    88 88.  .88"
"\n      dP   `88888P  88Y888P  dP dP    dP `8888P88"
"\n                    88                        .88"
"\n                    dP                    d8888P"
"\n"
"\ntcping is a ping over tcp connection.\n"
"\n"
"Examples:\n\
    1. ping over tcp with custom port 10 times\n\
        > tcping www.baidu.com 80\n\
    2. ping over tcp with custom port 5 times\n\
        > tcping -c 5 www.baidu.com 443\n\
\n\
USAGE: tcping [--count <count>] <ip> <port>\n\
\n\
ARGUMENTS:\n\
    <ip>                    The IP or Domain to tcping.\n\
    <port>                  The port to tcping.\n\
\n\
OPTIONS:\n\
    -c, --count <count>     The number of times to repeat 'tcping'. Default value is 10, Max value is 65535\n\
    -h, --help              Show help information.\n";
    
    NSString * verbose = [NSString stringWithFormat:v, VERSION];
    [ConsoleIO writeMessage:verbose to:OutputTypeStandard];
}

+ (void)printReulst:(BOOL)isFinish detail:(Tcping *)detail count:(NSUInteger)count lossCount:(NSUInteger)lossCount min:(NSNumber *)min max:(NSNumber *)max avge:(NSNumber *)avge {
    NSNumberFormatter * nf = [[NSNumberFormatter alloc] init];
    nf.numberStyle = NSNumberFormatterDecimalStyle;
    nf.maximumFractionDigits = 3;
    
    if (isFinish) {
        [ConsoleIO writeMessage:[NSString stringWithFormat:@"Ping statistics %@:%u\n", detail.domain, detail.port] to:OutputTypeStandard];
        [ConsoleIO writeMessage:[NSString stringWithFormat:@"    %lu probes sent.\n", count] to:OutputTypeStandard];
        [ConsoleIO writeMessage:[NSString stringWithFormat:@"    %lu successful, %lu failed.\n", count-lossCount, lossCount] to:OutputTypeStandard];
        [ConsoleIO writeMessage:@"Approximate trip times:\n" to:OutputTypeStandard];
        [ConsoleIO writeMessage:[NSString stringWithFormat:@"    Minimum = %@ms, Maximum = %@ms, Average = %@ms\n", [nf stringFromNumber:min], [nf stringFromNumber:max], [nf stringFromNumber:avge]] to:OutputTypeStandard];
    } else {
        if ([detail.host isEqualToString:detail.domain]) {
            [ConsoleIO writeMessage:[NSString stringWithFormat:@"%@:%u - Connected - %@ms\n", detail.host, detail.port, [nf stringFromNumber:[NSNumber numberWithDouble:detail.speed]]] to:OutputTypeStandard];
        } else {
            [ConsoleIO writeMessage:[NSString stringWithFormat:@"%@:%u has address: %@:%u - Connected - %@ms\n", detail.domain, detail.port, detail.host, detail.port, [nf stringFromNumber:[NSNumber numberWithDouble:detail.speed]]] to:OutputTypeStandard];
        }
    }
}
@end

File: tcping/ConsoleIO.swift
//
//  ConsoleIO.swift
//  tcping
//
//  Created by Youssef on 2020/3/25.
//  Copyright © 2020 Youssef. All rights reserved.
//

import Foundation

enum OutputType {
  case error
  case standard
}

class ConsoleIO {
    func writeMessage(_ message: String, to: OutputType = .standard) {
        switch to {
            case .standard:
                print("\(message)")
            case .error:
                fputs("Error: \(message)\n", stderr)
        }
    }
    
    func printUsage() {
        writeMessage("""
        tcping is a ping over tcp connection.

        Examples:
            1. ping over tcp with custom port 10 times
               > tcping www.baidu.com 80
            2. ping over tcp with custom port 5 times
               > tcping -c 5 www.baidu.com 443

        USAGE: tcping [--count <count>] <ip> <port>

        ARGUMENTS:
          <ip>                    The IP or Domain to tcping.
          <port>                  The port to tcping.

        OPTIONS:
          -c, --count <count>     The number of times to repeat 'tcping'. Default value is 10, Max value is 65535
          -h, --help              Show help information.
        """)
        NotificationCenter.default.post(name: NSNotification.Name("stop"), object: nil)
    }
    
    func printResult(isFinish: Bool = false, detail: tcping, count: UInt = 1, lossCount: UInt = 0, min: NSNumber = 0.0, max: NSNumber = 0.0, avge: NSNumber = 0.0) {
        let nf = NumberFormatter()
        nf.numberStyle = .decimal
        nf.maximumFractionDigits = 3

        if isFinish {
            self.writeMessage("Ping statistics \(detail.host):\(detail.port)")
            self.writeMessage("    \(count) probes sent.")
            self.writeMessage("    \(count-lossCount) successful, \(lossCount) failed.")
            self.writeMessage("Approximate trip times:")
            self.writeMessage("    Minimum = \(nf.string(from: min) ?? "")ms, Maximum = \(nf.string(from: max) ?? "")ms, Average = \(nf.string(from: avge) ?? "")ms")
        } else {
            if detail.host == detail.domain {
                self.writeMessage("\(detail.host):\(detail.port) - Connected - \(nf.string(from: NSNumber(value: detail.speed)) ?? "")ms")
            } else {
                self.writeMessage("\(detail.domain):\(detail.port) has address: (\(detail.host):\(detail.port)) - Connected - \(nf.string(from: NSNumber(value: detail.speed)) ?? "")ms")
            }
        }
    }
}
